Updating CPU code to match CUDA for torch.nn.InstanceNorm2d: Issue#139140

vishalgoyal316 opened the pull request:
Fixes #139140

vishalgoyal316 committed:
Fix for torch.nn.InstanceNorm2d throws mixed dtype error: Issue#139140

vishalgoyal316 committed:
Fix mixed dtype support in InstanceNorm2d (#139140)

InstanceNorm2d with track_running_stats=True fails when input dtype
differs from parameter dtype (e.g., float64 input, float32 params).

Fix: Convert parameters to input dtype before computation, then convert
updated stats back to original dtype. This ensures running stats are properly updated.

Optimizations: Cache dtype checks, fast path for same-dtype case.

Tests: Enhanced coverage for forward/backward with mixed dtypes.

vishalgoyal316 committed:
Simplify mixed dtype handling

- Remove fast/slow codepath split (single unified path)
- Only convert running_mean/var (weight/bias handled by batch_norm)
- Simplify tests: remove unnecessary loops and dead code
- Update error regex to match both CPU and CUDA messages

vishalgoyal316 commented:
Hello @isuruf, I have updated the PR, please review the code and provide your valuable feedback.

isuruf commented:
@pytorchbot rebase

ezyang commented:
@pytorchbot merge

vishalgoyal316 committed:
Fix InstanceNorm eval mode stats corruption Only update running_mean/running_var when use_input_stats=True to prevent stats corruption in eval mode.

vishalgoyal316 committed:
Fix instance_norm operation ordering for functionalization

Store at::alias() result in variable before computing updated stats
to ensure correct trace order: alias before view/mean operations.

Fixes test_functionalization.py::test_instance_norm failures.
No performance impact.

vishalgoyal316 committed:
[test] Support MPS in InstanceNorm mixed dtype tests

Use float16/float32 dtypes on MPS and float32/float64 on other devices
for mixed dtype testing. This allows the tests to run on MPS hardware
which doesn't support float64, eliminating the need for @expectedFailureMPS.

vishalgoyal316 committed:
[test] Fix MPS device detection for InstanceNorm mixed dtype tests

Fix _get_mixed_dtypes() to properly detect MPS device type from strings
like 'mps:0' by using device.split(':')[0] instead of direct string
comparison. This prevents attempting to create unsupported float64
tensors on MPS.

Fixes test_instancenorm_mixed_dtype_backward_mps failure.

vishalgoyal316 committed:
Update test/test_nn.py

Co-authored-by: Isuru Fernando <isuruf@gmail.com>

isuruf commented on test/test_nn.py:
```diff
@@ -8223,6 +8235,111 @@ def _test_LayerNorm_cpu_mixed_dtype(self, device, dtype):
+                         "Running stats should change between forward passes")
+        self.assertFalse(torch.allclose(running_means[1], running_means[2]),
+                         "Running stats should continue changing across passes")
+
```
It's better to use a `skipIfMPS` since the test will crash the python interpreter.

vishalgoyal316 committed:
Skip InstanceNorm mixed dtype test on MPS

Add @skipIfMPS decorator to test_instancenorm_mixed_dtype_backward.
MPS batch_norm doesn't support mixed dtypes and crashes with LLVM error.

vishalgoyal316 committed:
Fix InstanceNorm mixed running-stats dtype handling

Align running_mean/var casting with weight dtype when defined and add a CUDA
InstanceNorm1d mixed running-stats dtype test.

vishalgoyal316 committed:
Simplify InstanceNorm running-stats casting

Inline mixed-dtype running_mean/var casting into the repeat helper calls
without changing behavior.

vishalgoyal316 commented:
@pytorchbot merge
